### PR TITLE
Make CWallet::FundTransaction atomic

### PR DESCRIPTION
**Pull request by promag**
This PR fixes a race for `setLockedCoins` when `lockUnspents` is true. For instance, it should not be possible to use the same unspent in concurrent `fundrawtransaction` calls.

Now the `cs_main` and `cs_wallet` locks are held during `CreateTransaction` and `LockCoin`(s). Also added some style nits around the change.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -2593,17 +2593,19 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
...
     }
 
     coinControl.fAllowOtherInputs = true;
 
-    for (const CTxIn& txin : tx.vin)
+    for (const CTxIn& txin : tx.vin) {
         coinControl.Select(txin.prevout);
+    }
+
+    LOCK2(cs_main, cs_wallet);
```
Comment (ryanofsky) [inline src/wallet/wallet.cpp]: Would be good for a comment to explaining the lock. If I understand the PR description correctly, this is preventing concurrent fundraw transactions from spending the same outputs due to the delay between CreateTransaction and LockCoin below?

Comment (promag) [reply src/wallet/wallet.cpp]: Yes, or concurrent fundrawtransaction and lockunspent. The problem may not be the delay, but when one releases the lock and the other acquires, for instance.

I'll add the comment.

**Comment by Smallyen0329** [comment]
Comment on 4e4a774

**Comment by promag** [comment]
@Smallyen0329 mind to repeat the comment?

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (promag) [reply src/wallet/wallet.cpp]: Done.

Comment (TheBlueMatt) [reply src/wallet/wallet.cpp]: Can you drop cs_main from it? I don't think cs_main is required, only cs_wallet.

Comment (promag) [reply src/wallet/wallet.cpp]: Won't that change the lock order? `CreateTransaction` locks both cs_main and cs_wallet.

Comment (TheBlueMatt) [reply src/wallet/wallet.cpp]: Oops, indeed, yes, needs both.

**Review by ryanofsky — COMMENTED** [review]
utACK a3c92f4f459a385fccf13eb28f18b48009324fa5, looks good to me.

Just to complain a little, though, the change seems pretty drowned out by reformatting. In the future I'd maybe suggest doing this in another commit.

**Comment by promag** [comment]
@ryanofsky actually I was thinking removing the nits. But I can once I push again since atm needs squash.

**Comment by promag** [comment]
@ryanofsky split in 2 commits, same patch.

**Comment by laanwj** [comment]
LGTM, this is definitely better as two separate commits
utACK 03a5dc9

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -2592,18 +2592,18 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
 {
     std::vector<CRecipient> vecSend;
 
-    // Turn the txout set into a CRecipient vector
-    for (size_t idx = 0; idx < tx.vout.size(); idx++)
-    {
+    // Turn the txout set into a CRecipient vector.
+    for (size_t idx = 0; idx < tx.vout.size(); idx++) {
```
Comment (promag) [inline src/wallet/wallet.cpp]: Err, ++idx..

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -2613,31 +2613,29 @@ bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nC
...
+        // We don't have the normal Create/Commit cycle, and don't want to risk
+        // reusing change, so just remove the key from the keypool here.
         reservekey.KeepKey();
     }
 
-    // Copy output sizes from new transaction; they may have had the fee subtracted from them
-    for (unsigned int idx = 0; idx < tx.vout.size(); idx++)
+    // Copy output sizes from new transaction; they may have had the fee
+    // subtracted from them.
+    for (unsigned int idx = 0; idx < tx.vout.size(); idx++) {
```
Comment (promag) [inline src/wallet/wallet.cpp]: Err, ++idx..

**Pull request merged**